src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,366 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.util;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.Sets;
+import com.google.common.primitives.Ints;
+import org.apache.lucene.index.CheckIndex;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.store.SimpleFSDirectory;
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.ElasticsearchIllegalStateException;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.io.FileSystemUtils;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.ImmutableSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.env.ShardLock;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.settings.IndexSettings;
+import org.elasticsearch.index.shard.*;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ */
+public class MultiDataPathUpgrader {
+
+    private final NodeEnvironment nodeEnvironment;
+    private final ESLogger logger = Loggers.getLogger(getClass());
+
+    public MultiDataPathUpgrader(NodeEnvironment nodeEnvironment) {
+        this.nodeEnvironment = nodeEnvironment;
+    }
+
+    /**
+     * Upgrades the given shard Id from multiple shard paths into the given target path.
+     *
+     * @see #pickShardPath(org.elasticsearch.index.shard.ShardId)
+     */
+    public void upgrade(ShardId shard, ShardPath targetPath) throws IOException {
+        final Path[] paths = nodeEnvironment.shardPaths(shard); // custom data path doesn't need upgrading
+        if (isTargetPathConfigured(paths, targetPath) == false) {
+            throw new IllegalArgumentException("shard path must be one of the shards data paths");
+        }
+        if (paths.length == 1) {
+            logger.info("{} only one data path configured - skipping upgrade", shard);
+            return;
+        }
+        logger.info("{} upgrading multi data dir to {}", shard, targetPath.getDataPath());
+        final ShardStateMetaData loaded = ShardStateMetaData.FORMAT.loadLatestState(logger, paths);
+        if (loaded == null) {
+            throw new IllegalStateException("Can't upgrade shard without shard state");
```

Comment:
can we add the index name, shard id and the paths looked at as the exception? it's especially interesting because needsUpgrading checks for the state folder.  Also, do we care about nodes that have the state folders but no data in them? should we fail the node in that case? if so, may change the message to say "no shard state found in any of [FOLDERS], please check and remove them if possible".